### PR TITLE
Fix GetChocolatey so one can get chocolatey

### PR DESCRIPTION
**Pull request by mwrock**
I encountered a few problems with `GetChocolatey`. Here are the issues this fixes:

1. NullRefException when calling `Lets.GetChocolatey` when the agrs are parsed.
2. Calling `Lets.GetChocolatey` more than once in a single process raised an exception when setting the `ConstructorResolutionBehavior` on SimpleInjector since it can only be called once. I made some changes to `SimpleInjectorContainer` that ensures the initializer is only called once in the `lazy` initializer of `_container` and made that initializer private. So callers would instead call `SimpleInjectorContainer.Container` and the very first call initializes a new container. 
3. After fixing 2, adding items to the configuration option set was raising duplicate key errors. I changed the `ConfigurationOptions` to reset each time they are parsed.
4. There was no logging sent to stdout upon calling `run`. `GetChocolatey` now configures log4net but with no log file. Callers should be able to use `SetCustomLogger` if they do want file logging.

I added some integration tests for `GetChocolatey` and I ran all integration tests. 2 tests are failing but those same tests are failing in master as well so they appear to be unrelated.

### CODE REVIEW

**Comment by ferventcoder** [comment]
The tests failing in master are being fixed. I apparently wasn't quite ready to commit and push last night... :/

File: src/chocolatey/infrastructure.app/configuration/ConfigurationOptions.cs [hunk 1]
```diff
@@ -58,28 +56,25 @@ private static IConsole Console
...
+            }};
 
-            // add help only once
-            if (_optionSet.Count == 0)
-            {
-                _optionSet
-                    .Add("?|help|h",
-                         "Prints out the help menu.",
-                         option => configuration.HelpRequested = option != null);
-            }
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/configuration/ConfigurationOptions.cs]: What prompted this change? Does this lose options when coming through here twice?

Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/configuration/ConfigurationOptions.cs]: Yeah, this looks bad. You don't come through there twice b/c ConfigurationBuilder is your only call through, but a Console application comes through there twice. Please revert the changes to this file https://github.com/chocolatey/choco/blob/master/src/chocolatey/infrastructure.app/runners/ConsoleApplication.cs#L51

Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/configuration/ConfigurationOptions.cs]: Of note, test coverage is not as high as it needs to be to catch things like this yet.

File: src/chocolatey/GetChocolatey.cs [hunk 2]
```diff
@@ -58,7 +59,7 @@ public GetChocolatey()
 
         private void set_defaults()
         {
-            ConfigurationBuilder.set_up_configuration(null, _configuration, _fileSystem, _container.GetInstance<IXmlService>(), null);
+            ConfigurationBuilder.set_up_configuration(new List<string>(), _configuration, _fileSystem, _container.GetInstance<IXmlService>(), null);
```
Comment (ferventcoder) [inline src/chocolatey/GetChocolatey.cs]: Perf related change? Is there a hidden error when passing null?

File: src/chocolatey/GetChocolatey.cs [hunk 3]
```diff
@@ -49,8 +49,9 @@ public class GetChocolatey
 
         public GetChocolatey()
         {
+            Log4NetAppenderConfiguration.configure();
```
Comment (ferventcoder) [inline src/chocolatey/GetChocolatey.cs]: This locks folks into log4net.

Comment (ferventcoder) [inline src/chocolatey/GetChocolatey.cs]: I may have done that in the preview. I just don't want to build up technical debt though.

File: src/chocolatey/infrastructure/registration/SimpleInjectorContainer.cs [hunk 4]
```diff
@@ -24,7 +24,7 @@ namespace chocolatey.infrastructure.registration
     /// </summary>
     public static class SimpleInjectorContainer
     {
-        private static readonly Lazy<Container> _container = new Lazy<Container>(() => new Container());
+        private static readonly Lazy<Container> _container = new Lazy<Container>(initialize);
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure/registration/SimpleInjectorContainer.cs]: :+1:

File: src/chocolatey/GetChocolatey.cs (hunk 3 above)
Comment (ferventcoder) [inline src/chocolatey/GetChocolatey.cs]: I'm also not sure it actually sets the configuration active though. You would need a call to https://github.com/chocolatey/choco/blob/master/src/chocolatey/infrastructure/logging/Log4NetLog.cs#L18 to activate the logger.

Comment (ferventcoder) [inline src/chocolatey/GetChocolatey.cs]: My preference is for you to use a custom logger (like a console logger)?

File: src/chocolatey/GetChocolatey.cs (hunk 2 above)
Comment (ferventcoder) [inline src/chocolatey/GetChocolatey.cs]: Actually, I see that is 1 on your list.

File: src/chocolatey/infrastructure.app/configuration/ConfigurationOptions.cs (hunk 1 above)
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/configuration/ConfigurationOptions.cs]: I see this is number 3 in your list. We'll need a way to clear the option set explicitly. Please hook that up.

File: src/chocolatey/GetChocolatey.cs (hunk 3 above)
Comment (ferventcoder) [inline src/chocolatey/GetChocolatey.cs]: This may be okay.

Comment (ferventcoder) [inline src/chocolatey/GetChocolatey.cs]: Actually, if no logger is configured, how do you get this to work? 
It's set to NullLogger by default - https://github.com/chocolatey/choco/blob/master/src/chocolatey/infrastructure/logging/Log.cs#L27-L37

Comment (ferventcoder) [inline src/chocolatey/GetChocolatey.cs]: It's set by a call to Bootstrap.initialize - https://github.com/chocolatey/choco/blob/master/src/chocolatey/infrastructure/registration/Bootstrap.cs#L36

But then for some reason I am also erroneously setting it in ContainerBinding (where you get it) - https://github.com/chocolatey/choco/blob/master/src/chocolatey/infrastructure.app/registration/ContainerBinding.cs#L45

I need to fix that.

Comment (mwrock) [inline src/chocolatey/GetChocolatey.cs]: It gets set in the container registration: https://github.com/chocolatey/choco/blob/master/src/chocolatey/infrastructure.app/registration/ContainerBinding.cs#L45

Let me know if you want to change this. Seemed like this was a sensible default. It threw me off initially having `run` not emit anything.

Comment (mwrock) [inline src/chocolatey/GetChocolatey.cs]: weird. your last comment just popped up for me

Comment (mwrock) [inline src/chocolatey/GetChocolatey.cs]: Would it be better to call `Bootstrap.Initialize` instead of directly initializing the logger?

Comment (ferventcoder) [inline src/chocolatey/GetChocolatey.cs]: Yeah, I think that would be sufficient. :)

Comment (ferventcoder) [inline src/chocolatey/GetChocolatey.cs]: > Let me know if you want to change this.

I already fixed that. I just haven't pushed it yet.

Comment (ferventcoder) [inline src/chocolatey/GetChocolatey.cs]: And it's pushed.

File: src/chocolatey/GetChocolatey.cs (hunk 2 above)
Comment (ferventcoder) [inline src/chocolatey/GetChocolatey.cs]: :+1:

**Comment by mwrock** [comment]
updated PR

**Comment by ferventcoder** [comment]
LTE!

File: src/chocolatey/infrastructure.app/configuration/ConfigurationOptions.cs [hunk 5]
```diff
@@ -34,12 +34,17 @@ public static void initialize_with(Lazy<IConsole> console)
             _console = console;
         }
 
+        public static void reset_options()
+        {
+            _optionSet = new OptionSet();
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/configuration/ConfigurationOptions.cs]: Alternatively `_optionSet.Clear()`?

File: src/chocolatey.tests.integration/scenarios/GetChocolateyScenario.cs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            }
+
+            [Fact]
+            public void should_have_distinct_configurations()
+            {
+                chocolatey1.GetConfiguration().ShouldNotEqual(chocolatey2.GetConfiguration());
+            }
+        }
+    }
+}
```
Comment (ferventcoder) [inline src/chocolatey.tests.integration/scenarios/GetChocolateyScenario.cs]: This doesn't look to be an integration test. Why isn't this over in chocolatey.tests?

File: src/chocolatey/GetChocolatey.cs [hunk 7]
```diff
@@ -49,8 +49,10 @@ public class GetChocolatey
 
         public GetChocolatey()
         {
+            Log4NetAppenderConfiguration.configure();
+            Bootstrap.initialize();
```
Comment (ferventcoder) [inline src/chocolatey/GetChocolatey.cs]: :+1:

File: src/chocolatey/infrastructure/logging/Log4NetAppenderConfiguration.cs [hunk 8]
```diff
@@ -49,7 +50,9 @@ public static void configure(string outputDirectory)
             Stream xmlConfigStream = assembly.get_manifest_stream(resource);
 
             XmlConfigurator.Configure(xmlConfigStream);
-            set_file_appender(outputDirectory);
+
+            if(outputDirectory != null)
+                set_file_appender(outputDirectory);
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure/logging/Log4NetAppenderConfiguration.cs]: Style thing here, either set it all on the same line or introduce braces. I'm not a fan of implicit multi-line if statements.

**Comment by ferventcoder** [comment]
Okay so this is pretty much ready. Here are the next steps:

* Rebase against latest master
* Fix the couple of notes from above
* Squash your commits down
* Push the same branch with `-f`

Then we can move forward on this. :+1:

File: src/chocolatey/infrastructure/logging/Log4NetAppenderConfiguration.cs (hunk 8 above)
Comment (gep13) [inline src/chocolatey/infrastructure/logging/Log4NetAppenderConfiguration.cs]: This is something that will be enforced by StyleCop once I get it running.  StyleCop FTW! :-)

Comment (ferventcoder) [inline src/chocolatey/infrastructure/logging/Log4NetAppenderConfiguration.cs]: Oh sweet!

Comment (gep13) [inline src/chocolatey/infrastructure/logging/Log4NetAppenderConfiguration.cs]: Oh yes, StyleCop is very much your friend in this regard :-D

**Comment by mwrock** [comment]
cool. will fix up tonight.

File: src/chocolatey/infrastructure.app/configuration/ConfigurationOptions.cs (hunk 5 above)
Comment (mwrock) [inline src/chocolatey/infrastructure.app/configuration/ConfigurationOptions.cs]: yup. thats a better option. Missed that this inherits a collection.

File: src/chocolatey.tests.integration/scenarios/GetChocolateyScenario.cs (hunk 6 above)
Comment (mwrock) [inline src/chocolatey.tests.integration/scenarios/GetChocolateyScenario.cs]: ugh. it was going to be when I made it and then it wasn't. moving over.

File: src/chocolatey/infrastructure/logging/Log4NetAppenderConfiguration.cs (hunk 8 above)
Comment (mwrock) [inline src/chocolatey/infrastructure/logging/Log4NetAppenderConfiguration.cs]: yeah. even though I respectfully disagree with the style :) its nice when its cauht in the build.

Comment (gep13) [inline src/chocolatey/infrastructure/logging/Log4NetAppenderConfiguration.cs]: "To bracket, or not to bracket?" That is indeed the question! :-)

**Comment by mwrock** [comment]
Should be good to go. Let me know if anything doesn't look right.

**Comment by ferventcoder** [comment]
Almost all good. 

* Commit message doesn't follow https://github.com/chocolatey/choco/blob/master/CONTRIBUTING.md#prepare-commits
* Still expected at least 2-3 commits. 
  * One for log4netappender changes
  * One for SimpleInjector fixes
  * The last for fixes to GetChocolatey and ConfigurationOptions.

**Comment by ferventcoder** [comment]
I'll take it from here. :)

**Comment by ferventcoder** [comment]
Fixed up and merged into stable at https://github.com/chocolatey/choco/commit/33939dc33d8a92130baa2b261d839f1ff244233b

**Comment by mwrock** [comment]
Yeah I would have preferred that too. For some reason I thought you wanted it in one commit. Thanks for fixing.

**Comment by ferventcoder** [comment]
No worries.
